### DISCUSSION TITLE
Creating an edge with dynamic _from and _to?

### DISCUSSION THREAD

**Discussion by jerrytomato** [TomGrozev/arangox_ecto#70] [General]
What's the best way to define an edge collection that has no explicit _from and _to collection? I'm building an authorization system where the user can have access to many things through one edge collection "accesses" including a role. So far I have this, but I'm not able to do `use ArangoXEcto.Edge` without passing _from and _to.

```elixir
defmodule App.Schema.Accesses do
  use ArangoXEcto.Edge # Causes error because of no _from and _to

  import Ecto.Changeset

  schema "accesses" do
    field :role, Ecto.Enum, values: [:owner, :editor, :viewer], default: :owner

    edge_fields()
    timestamps(type: :utc_datetime)
  end

  @doc false
  def changeset(edge, attrs) do
    edges_changeset(edge, attrs)
    |> cast(attrs, [:role])
    |> validate_required(:role)
  end
end
```

**Comment by jerrytomato** [comment]
My hunch in the meantime if there is no better solution is to define another macro that I can then use from other modules for a "specific" _from and _to value. What do you think of something like this?

```elixir
defmodule App.Edge.Accesses do
  defmacro __using__(opts) do
    from = Keyword.fetch!(opts, :from)
    to = Keyword.fetch!(opts, :to)

    quote do
      use ArangoXEcto.Edge,
        from: unquote(from),
        to: unquote(to) 

      import Ecto.Changeset

      schema "accesses" do
        field :role, Ecto.Enum, values: [:owner, :editor, :viewer]

        edge_fields()
        timestamps(type: :utc_datetime)
      end

      @doc false
      def changeset(edge, attrs) do
        edges_changeset(edge, attrs)
        |> cast(attrs, [:role])
        |> validate_required(:role)
      end
    end
  end
end

defmodule App.Edge.UserTenantMembership do
  use App.Edge.Accesses,
    from: App.Users.User,
    to: App.Tenants.Tenant
end
```

**Reply by jerrytomato** [reply]
To me it seems like the primary reason for the ArangoXEcto.Edge module not support arbitrary from and to edges is because it tries to make use of belongs_to which needs an explicit schema with table. This is to support the existing Ecto associations. Is this correct?

**Reply by TomGrozev** [reply]
I don't see the what you are trying to achieve with that code snippet, it seems like it is the intended use of the edge module but you just reimplemented it, making it harder on yourself.

For edge relations you are correct, the issue here is Ecto. Since the multi type does not exist in relational databases. This is being tracked through https://github.com/TomGrozev/arangox_ecto/issues/39.

I have been stuck for ideas until recently and have thought of a possible way to implement it which I am going to try. Possibly leveraging some logic done in https://github.com/mathieuprog/polymorphic_embed

**Reply by jerrytomato** [reply]
Yeah that make sense. I realized for now, for my query purposes for this one specific thing it's better to just do it manually in a model like:

```
defmodule App.Authorizer do
  defmodule AuthorizationError do
    defexception message: "The user does not have the required privileges."
  end

  @roles [:viewer, :editor, :owner]
  @role_query """
   FOR a IN accesses
    FILTER a._from == @from && a._to == @to
    RETURN a.role
  """

  alias App.Repo

  def authorize_role!(from, to, required_role, do: block) when required_role in @roles do
    actual_role = role_for_model!(from, to)

    if role_index(actual_role) >= role_index(required_role) do
      block
    else
      raise AuthorizationError
    end
  end

  defp role_index(role) do
    @roles
    |> Enum.with_index()
    |> Keyword.fetch!(role)
  end

  defp role_for_model!(from, to) do
    Repo
    |> ArangoXEcto.aql_query(
      @role_query,
      from: from.__id__,
      to: to.__id__
    )
    |> case do
      {:ok, []} -> raise AuthorizationError
      {:ok, [first | _rest]} -> String.to_atom(first)
    end
  end
end
```

This is not too much of a problem. The issue comes in creating an item in an edge collection, since I need to create a changeset. For that I need a module that specifically has:

```
use ArangoXEcto.Edge,
    from: from,
    to: to
```
As I mentioned, for retrieval it makes sense a lot of the time to use a raw query. For creating or updating though, it's really advantageous to get use of the changesets. Any ideas for how these could work together a bit more seamlessly? I'm personally not concerned too much about supporting associations through Ecto's belongs_to etc. functions. Strange because it's kind of the opposite feeling I had before. I'd rather have something that better supports a generic changeset even if that means no automatic "loading" of the schema struct.

**Reply by TomGrozev** [reply]
This is now supported in version 2.0.
